aiohttp/resolver.py

```diff
@@ -30,11 +30,23 @@ def __init__(self) -> None:
 
         hosts = []
         for family, _, proto, _, address in infos:
-            hosts.append(
-                {'hostname': host,
-                 'host': address[0], 'port': address[1],
-                 'family': family, 'proto': proto,
-                 'flags': socket.AI_NUMERICHOST})
+            if family == socket.AF_INET6 and address[3]:
+                # This is essential for link-local IPv6 addresses.
+                # LL IPv6 is a VERY rare case. Strictly speaking, we should use
+                # getnameinfo() unconditionally, but performance makes sense.
+                host, _port = socket.getnameinfo(
+                    address, socket.NI_NUMERICHOST | socket.NI_NUMERICSERV)
+                port = int(_port)
+            else:
+                host, port = address[:2]
+            hosts.append({
+                'hostname': host,
+                'host': host,
+                'port': port,
```

Comment:
maybe normalize the port here instead of the if-block?
```suggestion
                'port': int(port),
```